app/javascript/controllers/setup_offline_controller.ts

```diff
@@ -0,0 +1,36 @@
+import { Controller } from "@hotwired/stimulus";
+import { wb } from "../serviceworker-companion.js";
+
+// Connects to data-controller="setup-offline"
+export default class extends Controller {
+  static targets = ["form", "password"];
+
+  declare readonly formTarget: HTMLFormElement;
+  declare readonly passwordTarget: HTMLInputElement;
+
+  connect() {
+    this.formTarget.addEventListener("submit", this.handleSubmit.bind(this));
+  }
+
+  async handleSubmit(event: Event) {
+    event.preventDefault();
+
+    // Pass in paths to CSS and JS to save for offline use, because document
+    // methods are not available in serviceWorker
+    const css = (
+      document.querySelector("link[rel=stylesheet]") as HTMLLinkElement
+    ).href;
+    const js = (document.querySelector("script[src]") as HTMLScriptElement).src;
+
+    await wb.messageSW({
+      type: "SAVE_CAMPAIGN_FOR_OFFLINE",
+      payload: {
+        password: this.passwordTarget.value,
+        campaignId: 1,
+        additionalItems: [css, js],
+      },
+    });
+
+    this.formTarget.submit();
```

Comment:
We're submitting the password after we initialise the cache with it, and download all the assets. If a password validation fails, on one hand it shouldn't matter much since we'll, in theory, download the data again when they try again to provide a proper password, BUT ... if they decide not to try the password again, for whatever reason, we're in a weird state. Personal data is downloaded to this device with a potentially weak password, which also won't be recoverable since the server has rejected it.

I think we should be more cautious with this. We could roll a fix forward, but we could also forget in which case we'll be susceptible to weird bugs, at the very least.

Reply:
In it's current incarnation, if the user submits an empty password it caches the data using that as an encryption key.